[website] MBT page (#1706)
informalsystems/quint: docs/pages/docs/model-based-testing.mdx

import { Callout, Steps } from 'nextra/components'

# Model-based Testing

> One of the most asked questions when it comes to writing specifications is "what about the code?". Or, in a more expressive problem statement:

Reviewer: ```suggestion
Quint is a specification language that shines in building complex distributed systems. In such projects there are two questions one needs to address:
1. Am I building the right system?
2. Am I building the system the right way?

When writing a specification in Quint, and using the tools "quint run --invariant..." and "quint verify --invariant." we can check that the design is right, that is, that the interactions of the components ensures that system-wide invariants are satisfied. At this point, big progress has been made: we can be sure that if we implement each of the components in a way that is consistent with the specification, then the system-wide invariants will also hold. We have reduced system-level correctness (potentially involving distributed and concurrent components) to making sure that each component in isolation is implemented according to the specification.

So, naturally, one of the most asked questions when it comes to writing specifications is "what about the code?". Or, in a more expressive problem statement:
```

> > How to bridge the gap between the specification and the implementation?

Reviewer: ```suggestion
> How to bridge the gap between the specification and the implementation?

... which is just another way of asking Question 2 from above.
```

Let's focus on more practical solutions:
  1. Model-based Testing (MBT)
  2. Trace validation

Both of these won't give you a proof that your code is correct, but they will make you more confident that the behavior between the model (specification) and the code (implementation) matches, even throughout changes to the code base (or the spec, which tends to change way less).

## MBT in one paragraph

Produce a bunch of traces (executions) from your model, which should be valid traces in your system. For example: Bob deposits $100 and then transfers $60 to Alice, and then Alice withdraws $20. A trace is just a JSON file with all the data you need. You then write a test driver to call the respective functions (`deposit("Bob", 100)`, `transfer("Bob", "Alice", 60)`, `withdraw("Alice", 20)`) in your implementation, and optionally check things between each call or at the very end (such as asserting a property or comparing the state). That's it, you use the model to generate test data (inputs and expectations) for you in a very efficient way.

| Data  | Expectations |
| :---  | :---         |
| Model | Model        |

> ## Trace validation in one paragraph

Reviewer: Is there more coming? We don't have a detailed section on this. We should at least mention that there is none, if there is none.

Author: I added this

<img width="1461" height="832" alt="image" src="https://github.com/user-attachments/assets/e77af7ee-48d2-4927-9a0c-b6259660596f" />

2. That the implementation function returns an error if and only if the trace also has an error registered for that state (i.e. "Could not withdraw because you don't have enough balance").
3. That some property holds (i.e. no balances are negative)

For the bank example, we check (2) by defining this auxiliary function:
```rust
fn compare_error(trace_error: Option<String>, app_error: Option<String>) {
    if trace_error.is_some() {
        assert!(
            app_error.is_some(),
            "Expected action to fail with error: {:?}, but it succeeded",
            trace_error
        );
        println!("Action failed as expected");
    } else {
        assert!(
            app_error.is_none(),
            "Expected action to succeed, but it failed with error: {:?}",
            app_error
        );
        println!("Action successful as expected");
    }
}
```

And then call it after each function invocation (all of them return a `Option<String>` like `deposit`):
```rust
"deposit_action" => {
    // ...
    let res = deposit(&mut bank_state, depositor, amount);
    compare_error(state.value.error.clone(), res)
}
```
</Steps>

### Running the complete example

> The repository includes a shell script file that will both generate the traces for you and run `cargo test`. The example includes a mismatch between model and implementation, where the implementation has a bug that allows a user to sell the same investment more than once, potentially draining funds from the bank. The model-based test we wrote identifies this issue. We can be much more confident that the model is correct as we can formally verify that this sort of draining never happens - and model-based testing transfers some of that confidence to the code. 

Reviewer: ```suggestion
The repository includes a shell script file that will both generate the traces for you and run `cargo test`. The example includes a mismatch between model and implementation, where the implementation has a bug that allows a user to sell the same investment more than once, potentially draining funds from the bank. The model-based test we wrote identifies this issue. 

Assuming that the specification is thoroughly analyzed and or verified, this in fact constitutes a bug in the implementation and not a protocol bug.
```
